### ISSUE TITLE
All files on /mnt/drive_letter are marked as executable

### ISSUE THREAD

**Issue by vladon** [Microsoft/WSL#936]
- All files in Windows accessed from Linux subsystem are marked with 'x' bit.
- Files such as .txt are not executable.
- Actual results (with terminal output if applicable)

```
$ ll /mnt/c/Windows/*.txt
-r-xr-xr-x 1 root root 0 Mar 15 07:49 /mnt/c/Windows/authtest.txt*
```
- Your Windows build number

14393.51

**Comment by rodrymbo** [comment]
I think that's been marked as a "feature" for months now. There's [documentation about why it was designed this way](https://blogs.msdn.microsoft.com/wsl/2016/06/15/wsl-file-system-support/). 

Basically, they only expect `/` (VolFS) to work like a linux filesystem. Anything under `/mnt` (DrvFS) is rwx for convenience. The underlying NTFS enforces access control. 

Execution (`x`) is problemmatic. If you don't have it, you can't execute anything stored in DrvFS -- which might make things more secure, but also leads to less flexibility, such as the ability to use a Windows text editor (which knows about Unix line-endings) to edit scripts, which one can't do in VolFS. 

The next iteration of this may also be on the backlog.

Apparently the way to change this is to find and support (or create) a suggestion on [uservoice](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo).

**Comment by Aspire1Inspire2** [comment]
This "all files as 'x'" is ugly under bash terminal. MINGW and cgwin don't do this.
I am still not using WSL as my default bash on windows due to this only reason.

**Comment by ncoghlan** [comment]
This is actually a functional issue for us, as it breaks `find`, and `file` under WSL.

**Comment by corydodt** [comment]
This is also a problem if you symlink files into your home directory to keep them in sync; for example, I have a standard practice to symlink .ssh (and other dotfiles) from a folder that is automatically backed up. SSH refuses to run this way because of the permissions on the linked files.

Some kind of workaround would be nice.

**Comment by b-hayes** [comment]
asside form uglyness this fucntionally breaks things as well such as when you are developing on ntfs but need wsl for deve tools and you may be using wsl for git.

Git will commit this X permission and now anyone who clones the repository is affected.
Any clones done in wsl will have a similar effect on any windows application that uses git.
From a windows application git will report that EVERY SINGLE FILE HAS CHNAGED. This will then cause a chain reaction of events all your git hooks and code valiadations script s etc etc eetc will now be processing files for no reason.

You can stop git tracking this but then shell scripts you have developed and tested will not be executable when cloned on a linux server.

There are sooo many problems caused by this. No file should be executable by default from WSL ever! (perhaps with the exceptions of .exe files for launching windows programs from wsl)

IMO there is no reason to make all file executable by default. No unix system ever does this when creating a file and neither does creating a new file from wsl, so why should files created in windows be different?

There is also no need to track all the permissions of files etc just use the linux default and only add +x in FS queries when '.exe,.com,.msi' files are concerned. Things that are obviously executable.

![image](https://user-images.githubusercontent.com/30540495/69838669-a75e4400-12a8-11ea-9e2c-7b9c0cd6a424.png)


Using permissions only matters from WSL and having  +x by default ONLY breaks things. It provides no benefit.

**Comment by jonathanpel** [comment]
I found this fix online.

Edit/create a WSL config file /etc/wsl.conf and add this inside it:

```
[automount]
enabled = true
options = "metadata,umask=027,fmask=117"
```

After you restart windows, files in your windows mount will not be flagged as "executable".
You will have to manually fix all the permissions for existing executable files on your drive since they will now be flagged as not-executable, but you should be fine after that.

I've only been using this for a day, but it seems to be working fine.

**Comment by b-hayes** [comment]
Thanks @jonathanpel I have actually tried this before but then you are not able to execute windows .exe files anymore even if you chmod +x them.

**Comment by jonathanpel** [comment]
You are right @b-hayes I didn't noticed that, but it's only true for apps that are in Windows "protected" folders. I have some executable files that are in my AppData folder or elsewhere and they run fine if I change the permission on them. I'm not able to change the permissions on files that are in Program Files or in System32 and such, so they can't be run from WSL.

For me it's a fine compromise since the few Windows executable I need to run from WSL are in my personal "Users" folder.
Clearly it's not a perfect solution, but I think it's a useful tweak while we wait for a more permanent solution.

**Comment by darkfader** [comment]
Perhaps a combination of
```
[automount]
enabled = false
mountFsTab = true
```
and
```
c:             /C             drvfs      rw,noatime,uid=1000,gid=1000,umask=027,fmask=117,metadata    0    0
//localhost/c$ /c             drvfs      ro,noatime,uid=1000,gid=1000,umask=027,fmask=007    0    0
/C/Users       /c/Users       none       bind   0    0
```
and
`export PATH="${PATH//\/C\//\/c\/}"`
This should make executable files (PATH) load via SMB (read-only is sufficient and no metadata needed) and other files in the normal way.

**Comment by fuhrmanator** [comment]
I also discovered this breaks python `nosetests` unless you (are willing to) provide `--exe` (read more at https://stackoverflow.com/a/6334161/1168342)

**Comment by garretwilson** [comment]
> Git will commit this X permission and now anyone who clones the repository is affected.

So in summary, as a general rule if a developer is using a default WSL installation (without the workaround in `/etc/wsl.conf`), they should be instructed always to make Git commits via Command/PowerShell/Git Bash and _never make Git commits from inside WSL_. (Or more precisely, never to add files to the Git index from within WSL.) Otherwise the repository will be littered with arbitrary source files marked as executable, based upon which developer happened to make the commit.

Is that a fair assessment?

**Comment by garretwilson** [comment]
> So your stance is EVERYONE should have to implement workarounds for the MAJORITY of cases rather than taking a sensible default position that benefits everyone?

I can't tell if your reply is directed at me, but since you quoted me I want to note that I'm not with Microsoft and I'm new to this ticket, and I was merely asking a question to confirm what Microsoft recommends as the implication of this decision. (Yes, I'm as incredulous as you are that this decision was made, but my post was merely to put down concisely in black-and-white what the decision and its implication was before I started complaining about it. If someone verifies it is true, which I'm still waiting on, then yes my reaction would be similar to yours.)

> Force dev to implement/use workarounds or avoid using Linux tools entirely defeats the purpose of having WSL imo.

Exactly. I was waiting to have that reaction until someone confirmed that this is indeed the case (which I infer to be true from reading the ticket, but since it is so ludicrous I wanted to make sure first).

So can someone confirm that my [summary above](https://github.com/microsoft/WSL/issues/936#issuecomment-1455154431) is correct, or am I mistaken in understanding the situation?

**Comment by b-hayes** [comment]
Sorry if my response seemed harsh.
It is a fair assessment however, if your only concern is git, then I would just type the following command in wsl...
`git config --global core.filemode false`

And then you can only commit a +x bit if you intentionally do so via another git command:
`git update-index --add --chmod=+x`

**Comment by garretwilson** [comment]
> if your only concern is git, then I would just type the following command in wsl...

So I have to main two separate `.gitconfig` files on the same machine? And each developer has to remember to make this change? And now adding an executable file in Git requires extra commands, and you have to remember to do it differently if you're at the PowerShell or WSL prompt?

That doesn't sound like "just" is appropriate in your sentence. It sounds like even that is a nightmare of maintainability.

> Sorry if my response seemed harsh.

If my assessment was correct (and it seems to have been), that the response wasn't harsh—I just wasn't sure why it was directed at me.

But the issue here is this incredibly "inconvenient" (for lack of a better word) approach on the part of WSL. Sheesh. WSL is so nice in general, but as usual, there's always a wrench being thrown in.

**Comment by b-hayes** [comment]
You would have to use that same git command on windows if you want to commit the +x bit regardless since the windows file system has no way to represent +x. The config setting also means nothing to windows because of this. If you dont want to use git inside wsl thats fine but I wouldnt force people not to when a simple config change will do.

**Comment by gojimmypi** [comment]
I came here looking for a reason why Visual Studio would think that all the bash script files in a fresh clone had changed, but there were no visible differences. In wsl, `git diff` and `git status` indicated there were no changes.

Recently I created a new `/etc/wsl.conf` file as none of my PRs included the executable permission on a new script file, even though it _appeared_ to have the executable flag. (I'm usign a shared file system in `/mnt/c/workspace`). The new contents of my `/etc/wsl.conf` file seemed to fix the executable bit in a commit:

```
[automount]
options = "metadata"
```

However a new, fresh clone would appear to have file changes. To resolve that I used this:

```
git config core.filemode false
```

I'm not sure this is the best solution, but it seems to be working. I'm open to suggestions or comments. Thanks
